### DISCUSSION TITLE
About rdkit morgan fincgerprint genration process

### DISCUSSION THREAD

**Discussion by Rojina99** [rdkit/rdkit#7705] [Q&A]
I have a setup in which I generate morgan fingerprint using rdkit as follows

`fpgen1 = rdFingerprintGenerator.GetMorganGenerator(radius=3, countSimulation=False,
                                                           includeChirality=False, fpSize=self.fp_size)
 state_rep_i = np.array([fpgen.GetFingerprint(mols).ToList()]).astype(np.uint8)
 `
 
 Here, mols is rdkit molecule object 
 
 `mols
<rdkit.Chem.rdchem.Mol object at 0x15b1b33e0>`

which when converted to smiles looks like

`Chem.MolToSmiles(mols)
'[Al].[OH2+]c1ccccc1'`

The smile string is not just a single molecule but concatenation of molecules Al and [OH2+]c1ccccc1.

In the documentation https://www.rdkit.org/docs/GettingStartedInPython.html#morgan-fingerprints they only mentioned example for single molecule. In original ECFP paper also https://pubs.acs.org/doi/full/10.1021/ci100050t?casa_token=hlIsTvVognMAAAAA%3AbV7TpXgYv0YXo8WaJv9jBvBFk7MHJ7H3QLyXsRODeuqvcIgcn1PWyTLvh8GIdZrGj2wI9Mh-6J28Wqo they mention the conversion process for single molecule only.

So, can anyone help me understand how fingerprint is generated for smile string with multiple molecules as in example. And also point me to the documentation regarding the same.

**Comment by greglandrum** [comment]
@Rojina99 Terminology is important here: from the RDKit's perspective what you get from `Chem.MolToSmiles()` is always a single molecule. In your case you have a single molecule that is composed of multiple fragments. 

The way the Morgan fingerprint algorithm works will lead you to get a fingerprint which is the union of the fingerprints of the two fragments, so the output fingerprint will have all bits set that are set in the fingerprints of either of the fragments

**Reply by Rojina99** [reply]
Thank you for correcting my terminology. So, in this case if we have more than one fragments it is possible that two completely different fragments can result in same fingerprint ? Like the for first molecule with two fragments if  first fragment has 1011 and second has 1001  resulting in 1011 and  for other molecule with two fragment if first  fragment having 1010 and second has 0001 resulting in 1011 ? 

Can you also if possible point me to the code of fingerprints ?

**Reply by thwelln** [reply]
Yes, it is possible with Morgan Fingerprints for two different  molecules two have the exact same bits enabled and therefore the same fingerprint as a whole. In Morgan fingerprints, it's even more drastic than that: Each individual fingerprint bit can theoretically be enabled by very different substructures. The reason for this is the way those fingerprints are generated, through a hashing function. Hashing functions can always have collisions where two totally different inputs produce the same output. Greg has actually written a blog post on that topic, you can find it here: https://rdkit.blogspot.com/2014/02/colliding-bits.html
Regarding the code for Fingerprint generation, it is probably written in C++, would that still be of help to you?
You might also want to read the paper on Extended-Connectivity Fingerprints, of which Morgan is a special case. Can you access this? https://pubs.acs.org/doi/10.1021/ci100050t

**Reply by Rojina99** [reply]
Thank you so much for your response.

`Regarding the code for Fingerprint generation, it is probably written in C++, would that still be of help to you?`

Yes

Thank you for the blog post, the data on bit collision is interesting. I am aware with ECFP paper. It is in my original question and thus aware of bit collision. But from the paper I got that if we have large enough fingerprint the amount of bit collision is less even though the data might be sparse. But here when we discussed union of two fragments it seems it is adding one more layer making it more probable to have collision. Suppose if I start with molecules which have large number of fragments then it is definitely going to collide.

**Reply by dehaenw** [reply]
When a molecule is composed of several fragments that are atom sized, it is indeed viable to do a collision attack with minimal resources. see this repo in which i did this (just for fun) https://github.com/dehaenw/morcata

In my opinion if the number of atoms and bonds in the molecule is comparable, the one that has several fragments will probably actually be slightly more sparse because it will miss a few higher radius circular environments on the fragment-fragment interface in the connected molecule. in this sense you could say your chance of a bit collision will actually go down when you have several fragments, because you have less bits.

**Reply by bp-kelley** [reply]
Fingerprint collision is another example of the birthday problem.  

https://en.wikipedia.org/wiki/Birthday_problem

The more invariants that are hashed/folded into a fixed length bit vector, the more likely you are to have a collision up to a probability of 1.  I was tangentially involved in looking at this far too long ago when I was at Bioreason

https://web.archive.org/web/20031124184959id_/http://home.postech.ac.kr:80/~arang/JCICS/2001/No.1/134.pdf

**Reply by bp-kelley** [reply]
Here a "fragment" is really "another compound" and increases the odds of collisions.

**Reply by Rojina99** [reply]
@bp-kelley Thank you so much. Your explanation is very helpful. It also validates my original concern regarding collision.  The work you have posted is interesting. Can you suggest any direction I can explore to use with fingerprint with which we can decrease the collision ?
